src/store.rs

```diff
@@ -0,0 +1,196 @@
+
+        for output in job.outputs.iter().sorted() {
+            let source = PathBuf::from(output.as_str());
+
+            ///////////////////////////////////////////////////
+            // Step 1: Validate that the path we got is safe //
+            ///////////////////////////////////////////////////
+            for component in source.components() {
+                match component {
+                    Component::Prefix(_) | Component::RootDir =>  anyhow::bail!(
+                        "Got `{}` as an output, but absolute paths are not allowed as outputs. Remove the absolute prefix to fix this!",
+                        source.display(),
+                    ),
+
+                    Component::ParentDir => anyhow::bail!(
+                        "Got `{}` as an output, but relative paths containing `..` are not allowed as inputs. Remove the `..` to fix this!",
+                        source.display(),
+                    ),
+
+                    Component::CurDir | Component::Normal(_) => (),
+                };
+            }
+
+            ////////////////////////////////////////////////////////////////
+            // Step 2: add the file content to the content-addressed hash //
+            ////////////////////////////////////////////////////////////////
+            hasher.update(output.as_bytes());
+
+            let mut file = File::open(&workspace.join(output.as_str())).with_context(|| {
+                format!(
+                    "couldn't open `{}` for hashing. Did the build produce it?",
+                    output
+                )
+            })?;
+
+            // TODO: docs for Blake3 say that a 16 KiB buffer is the most
+            // efficient (for SIMD reasons), but `std::io::copy` uses an 8KiB
+            // buffer. Gonna have to do this by hand at some point to take
+            // advantage of the algorithm's designed speed.
+            std::io::copy(&mut file, &mut hasher).with_context(|| {
+                format!("could not read `{}` to calculate hash", source.display())
+            })?;
+
+            ///////////////////////////////////////////////
+            // Step 3: make sure any parent paths exist  //
+            ///////////////////////////////////////////////
+            let mut ancestors: Vec<&Path> = source.ancestors().skip(1).collect();
+            ancestors.pop(); // we've made sure this is relative, so the first item is ""
+            ancestors.reverse(); // go `[a, a/b, a/b/c]` instead of `[a/b/c, a/b, a]`
+
+            for ancestor_path in ancestors {
+                let ancestor = ancestor_path.to_path_buf();
+
+                if output_dirs.contains(&ancestor) {
+                    continue;
+                }
+
+                log::trace!(
+                    "creating {} in {}",
+                    &ancestor.display(),
+                    temp.path().display()
+                );
+                std::fs::create_dir(temp.path().join(&ancestor)).with_context(|| {
```

Comment:
that might simplify the code some, yeah, but at a performance cost:

- we'd be doing a filesystem read for each item in the ancestor chain for every item in the outputs to check if we need to make the directory
- we'd have to walk all the output paths again later to make the directories read-only after all the files are situated within them

That approach would save us some memory and a little bit of complexity, but make disk performance much worse. I could still see it being *ok* though, since ideally we will be avoiding rebuilds (and the subsequent workspace consumptions) as much as possible. The exception would be on the first build, where we'll be rebuilding everything.

Reply:
> we'd be doing a filesystem read for each item in the ancestor chain

Not exactly. create_dir_all, starts at the child node and works backwards towards the ancestors as necessary. So to create `a/b/c/d/e` where `a/b` already exists, it will:
1. try and fail to create `a/b/c/d/e`
1. try and fail to create `a/b/c/d`
1. succeed at creating `a/b/c`
1. succeed at creating `a/b/c/d`
1. succeed at creating `a/b/c/d/e`

Not sure the cost of failing to create a dir because it's parent doesn't exist, but assuming that is cheap, this should be fast.

> we'd have to walk all the output paths again later to make the directories read-only after all the files are situated within them

True, the cost here is probably fs dependent. I would assume not marking anything read only and then later marking everything read only would have low cost. 1, most things should be in fs cache. 2, you would only be traversing metadata that should be stored pretty compactly. Also, I would assume the cost to hash and copy files would overshadow all of this anyway.


Just to clarify, in this scenario, either everything or nothing should exist, correct? Since either we ran before and have all the outputs, or we haven't and need to build all directories. If this is correct, your most efficient option would be to process all ancestors of all files at once and just do a de-duplication of the list of ancestors.

Reply:
> Not exactly. create_dir_all, starts at the child node and works backwards towards the ancestors as necessary.

Ah, that's on me for assuming that `create_dir_all` would start from the root. Thanks for clarifying.

> Also, I would assume the cost to hash and copy files would overshadow all of this anyway.

Hashing, sure, but we're renaming the files instead of copying them since the original bytes would be reaped anyway when the build workspace was deleted.

> Just to clarify, in this scenario, either everything or nothing should exist, correct?

Yeah, that's right.

> If this is correct, your most efficient option would be to process all ancestors of all files at once and just do a de-duplication of the list of ancestors.

Let me rephrase this to make sure I understand what you're suggesting:

1. calculate the hash based on the files
2. exit early if we already have the hash
3. do the filesystem work, deduplicating ancestors

I ask because we already are effectively deduplicating ancestors—we're leaning heavily on the fact that the output directory won't exist in advance, so we don't need to check if some parent directory exists in advance of creating it. I wonder about combining our suggestions, such that the "do the filesystem work" above looks like:

1. create the temporary collection directory
2. create all the directories
3. move all the files, marking them as read-only
4. mark all the directories as read-only
5. move the collection directory into the store

(I want to insist on steps 1 and 5 for safety, even though they'll add just a little overhead. It'd be really bad to have a corrupt path in the CAS!)

Anyway, I think this could end up much nicer, both from an efficiency perspective and an ease-of-understanding one. We could reuse the approach for create-move-mark here, or use some other kind of filesystem walking like `create_dir_all`. (I think the deciding factor maybe should be approachability instead of speed here, although with some reasonable documentation of assumptions I think the current approach could be fairly understandable.)

Reply:
oh also, this maybe should be a separate refactor in a new PR! The current approach works OK—this would get that last little bit of performance out and make it easier to understand, but I don't think it makes sense to block the rest of the outstanding PRs on that work.

Reply:
Yeah, I am fine with either solution, just feel there are ways to simplify the readability here. Anyway, not needed for this PR for sure.

Reply:
yeah. I think I do want to change it. Might as well apply the guiding principles (understandable, then approachable, then fast) to the code as well as the system's behavior. Thank you for this excellent review!